Pedal-Intelligence/saypi-userscript
# Add Error Handling States to SayPi XState Machine

This PR introduces several error states to the SayPi XState machine to enhance error handling and user experience. Specifically, a parent `errors` state has been added, encapsulating various error-related child states like `transcribeFailed` and `micError`.

#### Changes

1. **Error States**: Added a parent `errors` state with child states:
    - `transcribeFailed`: Handles transcription API failures.
    - `micError`: Handles microphone errors such as no audio input detected.
    
2. **Transitions**: Added new transitions to error states from `transcribing` state. These transitions are triggered by events like `saypi:transcribeFailed` and `saypi:transcribedEmpty`.

3. **Visual Indicators**: Added animations and notifications as visual indicators for error states.
    - `transcribeFailed`: Shows an error animation.
    - `micError`: Displays a muted microphone notification icon.
  
4. **Auto Recovery**: Implemented automatic recovery from error states back to the `idle` state after 10 seconds.

5. **Notification Handling**: Updated `ButtonModule.js` to handle the display and dismissal of notifications.

#### Screenshots

Empty transcription error (mobile):
<img width="361" alt="saypi muted mic mobile" src="https://github.com/Pedal-Intelligence/saypi-userscript/assets/16578183/591624b8-9992-4fd4-9d1c-cfee45bf70fd">

Transcribe API error (mobile):
<img width="364" alt="saypi transcribe error mobile" src="https://github.com/Pedal-Intelligence/saypi-userscript/assets/16578183/f8b9e266-35af-4a31-bc21-1df260c9fb5a">

#### How to Test

1. Checkout to this branch.
2. Run the application.
3. Try various error scenarios to ensure the machine transitions to the appropriate error states.
4. Ensure visual indicators such as animations and notifications appear as expected.
5. Verify that the machine returns to the `idle` state after 10 seconds in an error state.
